src/orquestra/quantum/api/wavefunction_simulator.py

```diff
@@ -0,0 +1,107 @@
+################################################################################
+# © Copyright 2022 Zapata Computing Inc.
+################################################################################
+
+from abc import abstractmethod
+from typing import Optional, Protocol
+
+import numpy as np
+
+from ..circuits import Circuit, GateOperation, Operation, split_circuit
+from ..distributions import (
+    MeasurementOutcomeDistribution,
+    create_bitstring_distribution_from_probability_distribution,
+)
+from ..measurements import Measurements
+from ..operators import PauliRepresentation, get_expectation_value
+from ..wavefunction import Wavefunction, sample_from_wavefunction
+from ..typing import StateVector
+
+from .circuit_runner import BaseCircuitRunner, CircuitRunner
+
+
+class WavefunctionSimulator(CircuitRunner, Protocol):
+    def get_wavefunction(
+        self, circuit: Circuit, initial_state: Optional[StateVector] = None
+    ) -> Wavefunction:
+        pass
+
+    def get_exact_expectation_values(
+        self, circuit: Circuit, operator: PauliRepresentation
+    ) -> complex:
+        pass
+
+
+class BaseWavefunctionSimulator(BaseCircuitRunner, WavefunctionSimulator):
+    def __init__(self, *, seed: Optional[int] = None):
+        super().__init__()
+        self.seed = seed
+
+    def run_and_measure(self, circuit: Circuit, n_samples: int) -> Measurements:
+        if n_samples <= 0:
```

Comment:
This function is not counting the number of times a circuit was executed.

Reply:
I think it does, but because it is counted in `get_wavefunction` method which is used by this function :slightly_smiling_face: